user_topics: Add a new endpoint to update visibility_policy.

prakhar1144 opened the pull request:
This commit adds a new endpoint, `POST /user_topics` which is used to update the **visibility_policy** of a user topic row.

[CZO discussion](https://chat.zulip.org/#narrow/stream/378-api-design/topic/muted.20topics/near/1531109)

**Remaining decisions:**
- [x]  Allow you to remove a topic visibility policy even if you're no longer subscribed to the stream containing the topic? -- [CZO discussion](https://chat.zulip.org/#narrow/stream/3-backend/topic/Unmute.20Topics/near/1537641) -- I would need to do some tweaking to the PR after a decision is made. ([Decision](https://chat.zulip.org/#narrow/stream/3-backend/topic/Unmute.20Topics/near/1539950))
- [x] Should [update_muted_topic](https://github.com/zulip/zulip/blob/main/zerver/views/user_topics.py#L64) internally call the new view? -- (Possibly NO, because request variables will be verified twice and we're not getting any real advantage.)
- [x] Should we repeat [test_add_muted_topic](https://github.com/zulip/zulip/blob/main/zerver/tests/test_user_topics.py#L87) with the new URL? -- or we handle such cases differently when deprecating some URLs. -- (I have already added tests for the UNMUTED case.)
- [x] What types of error responses do we document? -- Currently, [/api/mute-topic](https://zulip.com/api/mute-topic#response) doesn't document error responses, but it is raised when we provide an invalid `stream_name` or `stream_id`.

<!-- Describe your pull request here.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
FYI: @timabbott @abhijeetbodas2001 

Questions in the PR description needs attention.

abhijeetbodas2001 commented:
> Should [update_muted_topic](https://github.com/zulip/zulip/blob/main/zerver/views/user_topics.py#L64) internally call the new view? -- (Possibly NO, because request variables will be verified twice and we're not getting any real advantage.)

I don't think so. The `do_set_user_topic_visibility_policy` function we extracted out seems DRY enough to me :)

> Should we repeat [test_add_muted_topic](https://github.com/zulip/zulip/blob/main/zerver/tests/test_user_topics.py#L87) with the new URL? -- or we handle such cases differently when deprecating some URLs. -- (I have already added tests for the UNMUTED case.)

The new endpoint should definitely be tested. It would be best to test `do_set_user_topic_visibility_policy` separately for all the error handling, and test the views only a few times to verify whatever logic is not present in `do_set_user_topic_visibility_policy`.

> What types of error responses do we document?

I'm not sure. Maybe raise this question in `#api design`.


CC @gnprice

timabbott commented:
> The new endpoint should definitely be tested. It would be best to test do_set_user_topic_visibility_policy separately for all the error handling, and test the views only a few times to verify whatever logic is not present in do_set_user_topic_visibility_policy.

I'd actually prefer all the new tests use the views. See https://zulip.readthedocs.io/en/latest/testing/philosophy.html#integration-testing-or-unit-testing for some of the reasoning, but basically it's almost always better to end-to-end verify behavior than just verifying behavior of internal functions, because it's easy for the internal function to be correct but still have incorrect overall output. 

For documenting error responses, generally our policy is to only document those error responses that we expect a client to want to do something special other than "display the error message" in response to.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,11 @@ format used by the Zulip server that they are interacting with.
+**Feature level 170**
+
+* [`POST /user_topics`](/api/update-user-topic):
+  Added a new endpoint to update the visibility policy of a topic.
```
Technically, this endpoint updates the user's preferences for a topic -- the distinction is important in that in the future we may have other topic properties than just visibility policy managed ny this endpoint.

prakhar1144 replied:
Earlier I thought of documenting only those things which we currently offer and update the docs later on when we start supporting other things.

But the suggested text makes more sense. (It explains the future scope and the current features we offer.)

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8261,6 +8266,89 @@ paths:
+  /user_topics:
+    post:
+      operationId: update-user-topic
+      summary: Update visibility policy of a topic
```
Same naming comment here and in the sidebar index -- "Update personal preferences for a topic" is a better summary; in the longer-form description I'd extend the sentence to say "such as the topic's visibility policy, which is used to implement [muted topics](/help/mute-a-topic) and related features."

prakhar1144 replied:
Updated the docs.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8261,6 +8266,89 @@ paths:
+        in a muted stream that the current user is subscribed to.
+        MUTED topics are displayed faded in the Zulip UI and are not included
+        in the user's unread count totals. An UNMUTED topic in a muted stream
+        behaves as an unmuted topic in an unmuted stream.
```
The `visibility_policy` documentation belongs in the description of the field itself, not in the endpoint's description, which should just explain that this modifies for the visibility policy for the single stream and topic pair indicated by the parameters.

prakhar1144 replied:
Makes sense -- updated!

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8261,6 +8266,89 @@ paths:
+
+            Clients must provide either `stream_name` or `stream_id` as a parameter
+            to this endpoint, but not both. Clients should use `stream_id`
+            instead of the `stream_name` parameter when possible.
```
Do we need to support `stream_name`? I think we have it on the other endpoint for backwards-compatibility, but I'd prefer to just not offer it here given that it's a new endpoint.

prakhar1144 replied:
Cool, removed `stream_name` -- `stream_id` is now a required parameter.

timabbott commented on zerver/tests/test_user_topics.py:
```diff
@@ -283,3 +283,149 @@ def set_topic_visibility_for_user(user: UserProfile, visibility_policy: int) ->
         )[0].last_updated
         self.assertEqual(cordelia_date_unmuted, date_unmuted)
+
+    @mock.patch("zerver.actions.user_topics.send_event")
```
Why do we mock this here? The `with tornado_redirected_to_list` function is our standard way to skip Tornado logic and capture events... (I guess I see we use this pattern in `test_message_edit`, but I think we should plan to replace that with our standard API). 

I also wonder if `tornado_redirected_to_list` should be renamed to something like `with capture_send_event_calls()` or something -- that might be a good #backend topic for you to start. (I just had the thought that the name may not be super self-explanatory)

prakhar1144 replied:
> Why do we mock this here?

To verify if events are sent properly. Currently in `MutedTopicsTests`, we verify the event's content in [this line](https://github.com/zulip/zulip/blob/main/zerver/tests/test_user_topics.py#L109). We should instead verify if it works end-to-end, right?
(Let me know, if this approach looks good, then i will update the existing code)

> 'with tornado_redirected_to_list' function is our standard way to skip Tornado logic and capture events

used this. (I will update `test_message_edit` to follow this pattern in a separate PR)

> I also wonder if tornado_redirected_to_list should be renamed to something like with capture_send_event_calls() or something -- that might be a good #backend topic for you to start

Sounds better. (I will open a topic)

timabbott replied:
Yeah, that's correct, sounds good!

timabbott commented on zerver/tests/test_user_topics.py:
```diff
@@ -283,3 +283,149 @@ def set_topic_visibility_for_user(user: UserProfile, visibility_policy: int) ->
+            },
+        ]
+
+        mock_date_unmuted = datetime(2020, 1, 1, tzinfo=timezone.utc).timestamp()
```
We recently added the `time_machine` module as a dependency in `test_invite.py`, so you could consider using that if it feels nicer than mocking (I've not played with it much).

prakhar1144 replied:
Used `time_machine` module.

(interesting module with a lot of added capabilities :))

abhijeetbodas2001 commented:
> I'd actually prefer all the new tests use the views. See https://zulip.readthedocs.io/en/latest/testing/philosophy.html#integration-testing-or-unit-testing for some of the reasoning, but basically it's almost always better to end-to-end verify behavior than just verifying behavior of internal functions, because it's easy for the internal function to be correct but still have incorrect overall output.

Thanks for the link! This makes sense.
Above, I was mostly thinking about avoiding test duplication. The `do_` function raises all of our errors, and the views are thin wrappers around that, with little validation of their own (apart from REQ). I guess we should avoid verifying all the errors for the two views independently (given the place of raising the errors is the same for both)?

The argument for writting out the set of tests for the new endpoint from scratch could be that it would be easy to just delete the test for the old endpoint once we get rid of the old endpoint?

prakhar1144 commented:
@abhijeetbodas2001 

Just to be precise, the `do_` function doesn't raise any error (It just `logging.info` in couple of cases). :)
All the errors raised are due to [`check_for_exactly_one_stream_arg`](https://github.com/zulip/zulip/blob/main/zerver/views/user_topics.py#L72),  [`access_stream_by_id`](https://github.com/zulip/zulip/blob/main/zerver/views/user_topics.py#L33) and [`access_stream_by_name`](https://github.com/zulip/zulip/blob/main/zerver/views/user_topics.py#L30).

The new view [won't contain](https://github.com/zulip/zulip/pull/24943#discussion_r1155146497) `stream_name`. So, `check_for_exactly_one_stream_arg` and `access_stream_by_name` will be removed.

Also, in the new test our [payload data](https://github.com/zulip/zulip/blob/main/zerver/tests/test_user_topics.py#L95) will change. i.e. we no longer have `stream_name` and `op = add/remove` replaced with `visibility_policy`.

So, i think creating the 'set of tests for the new endpoint from scratch' would be a cleaner approach (including the fact it would be easy to just delete the test for the old endpoint once we get rid of the old endpoint)

We should maybe rename the existing test class as `MutedTopicsTestsDeprecated` with an added comment about deprecation?

and the new test class named as `MutedTopicsTests`.

abhijeetbodas2001 commented:
> Just to be precise, the do_ function doesn't raise any error (It just logging.info in couple of cases). :)
All the errors raised are due to [check_for_exactly_one_stream_arg](https://github.com/zulip/zulip/blob/main/zerver/views/user_topics.py#L72), [access_stream_by_id](https://github.com/zulip/zulip/blob/main/zerver/views/user_topics.py#L33) and [access_stream_by_name](https://github.com/zulip/zulip/blob/main/zerver/views/user_topics.py#L30).

Yeah, I meant stuff internal to that call stack (I also had the DB integrity errors in mind, but it seems I had forgotten that we had decided on returning success for duplicate requests :))

> We should maybe rename the existing test class as MutedTopicsTestsDeprecated with an added comment about deprecation?

That makes sense to me 👍

prakhar1144 commented:
This is ready for another look.

Updates in the recent push:
1. Addressed the review comments (I have replied to the respective review comments)
2. Added commits to enforce end-to-end test instead of calling internal functions. (check the first and last commit)

Needs attention: [Question (not a blocker)](https://github.com/zulip/zulip/pull/24943#discussion_r1155491534)

FYI: @abhijeetbodas2001 @timabbott

prakhar1144 commented:
This is ready for another look.

Updates in the recent push:
1. Addressed the review comments (I have replied to the respective review comments)
2. Added commits to enforce end-to-end test instead of calling internal functions. (check the first and last commit)

Needs attention: [Question (not a blocker)](https://github.com/zulip/zulip/pull/24943#discussion_r1155491534)

FYI: @abhijeetbodas2001 @timabbott

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -8261,6 +8266,71 @@ paths:
+
+            MUTED topics are displayed faded in the Zulip UI, are not included
+            in the user's unread count totals, and the user doesn't receive any
+            notifications. An UNMUTED topic in a muted stream behaves as an
+            unmuted topic in an unmuted stream.
```
```suggestion
            notifications.

            An UNMUTED topic will remain visible even if the stream
            is muted. In a stream that is not muted, a policy of
            UNMUTED has the same effect as INHERIT.
```

timabbott replied:
change lgtm!

prakhar1144 replied:
Updated

timabbott commented on zerver/openapi/python_examples.py:
```diff
@@ -728,6 +728,44 @@ def toggle_mute_topic(client: Client) -> None:
+    request = {
+        "stream_id": stream_id,
+        "topic": "dinner",
+        "visibility_policy": 1,
```
This should be a follow-up, but once this is done, we should add an enum class to the `python-zulip-api` API bindings project so that this code can use a declarative name for the value imported from the module -- it'll make this example feel a lot nicer to read.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -8261,6 +8266,71 @@ paths:
+        The `/users/me/subscriptions/muted_topics` endpoint, which used to only
+        mute or unmute a topic, is now deprecated.
+        It exists for backwards compatibility and will be removed once clients
+        have migrated to use this endpoint.
```
I propose replacing everything after `170).` with the following, to avoid "unmute a topic" as a phrase, just to avoid confusion with the new feature that endpoint didn't support.

"Previously, toggling the muting state for a topic was managed by the
`/users/me/subscriptions/muted_topics` endpoint, which this endpoint
is intended to replace."

(The other motivation here is to avoid having text here that we'll need to remember to update when we do remove the old endpoint).

prakhar1144 replied:
Updated

timabbott commented on zerver/tests/test_user_topics.py:
```diff
@@ -229,6 +232,225 @@ def test_muted_topic_remove_invalid(self) -> None:
+                "topic": "test TOPIC",
+                "visibility_policy": visibility_policy,
+            }
+            with time_machine.travel(datetime(2020, 1, 1, tzinfo=timezone.utc), tick=False):
```
Let's use the `date_muted` variable in this line -- it's best not to repeat long timestamps like this one.

prakhar1144 replied:
Makes sense! -- updated.

timabbott commented:
I posted a handful of small comments, but otherwise this lgtm, please PM me when you have this updated so I can try to integrate it before we need to renumber the feature level.

prakhar1144 commented:
Updates in the recent push:
* Addressed the review comments.

timabbott commented:
Merged, thanks for all the work on this getting this nice and clean @prakhar1144!
